geo/geo.go

```diff
@@ -0,0 +1,243 @@
+package geo
+
+import (
+	"encoding/json"
+	"errors"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+)
+
+type GeoProvider interface {
+	GetLocation(ip string) (*Location, error)
+	FileName() string
+	DownloadUrl() string
+	LastUpdate() time.Time
+	NeedUpdate() bool
+	MaybeUpdate() error
+	Ready() bool
+	SetReady(bool)
+	SetFileName(string)
+}
+
+type Location struct {
+	IP          string `json:"ip"`
+	CountryCode string `json:"countryCode"`
+	Country     string `json:"country"`
+	City        string `json:"city"`
+}
+
+var emptyLocation = Location{}
+
+func (l *Location) Empty() bool {
+	if l == nil {
+		return true
+	}
+	return *l == emptyLocation
+}
+
+func (l *Location) Match(location *Location) bool {
+	if len(l.IP) > 0 && location.IP == l.IP {
+		return true
+	}
+	if len(l.CountryCode) > 0 && strings.ToLower(location.CountryCode) == strings.ToLower(l.CountryCode) {
+		return true
+	}
+	return false
+}
+
+type IPFilter struct {
+	DBPath    string        `json:"dbpath"`
+	Download  bool          `json:"download"`
```

Comment:
1. Suggest rename to 'DownloadDB'
2. `DBPath` and `Download` should be configured per tuna, not per service (per IPFilter), same for `Providers`.